src/cargo/core/resolver/mod.rs

```diff
@@ -298,109 +244,213 @@ fn activate(mut cx: Box<Context>,
+                self.next_index += 1;
+                Some((index, val))
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+struct DepsFrame {
+    parent: Rc<Summary>,
+    remaining_siblings: RcVecIter<DepInfo>,
+    platform: Option<String>,
+    id: PackageId,
+}
+type RemainingDeps = Vec<DepsFrame>;
+
+struct BacktrackFrame {
+    context_backup: Context,
+    deps_backup: RemainingDeps,
+    remaining_candidates: RcVecIter<Rc<Summary>>,
+    // For building an activation error:
+    parent: Rc<Summary>,
+    cur: usize,
+    dep: Rc<Dependency>,
+    prev_active: Vec<Rc<Summary>>,
+    all_candidates: Vec<Rc<Summary>>,
+}
+
+/// Recursively activates the dependencies for `top`, in depth-first order,
+/// backtracking across possible candidates for each dependency as necessary.
 ///
 /// If all dependencies can be activated and resolved to a version in the
-/// dependency graph the `finished` callback is invoked with the current state
-/// of the world.
-fn activate_deps<'a>(cx: Box<Context>,
-                     registry: &mut Registry,
-                     parent: &Summary,
-                     platform: Option<&'a str>,
-                     mut deps: slice::Iter<'a, DepInfo>,
-                     cur: usize,
-                     finished: &mut FnMut(Box<Context>, &mut Registry) -> ResolveResult)
-                     -> ResolveResult {
-    let &(dep, ref candidates, ref features) = match deps.next() {
-        Some(info) => info,
-        None => return finished(cx, registry),
-    };
+/// dependency graph, cx.resolve is returned.
+fn activate_deps_loop(mut cx: Context,
+                      registry: &mut Registry,
+                      top: Rc<Summary>,
+                      top_method: &Method) -> CargoResult<Resolve> {
+    let mut backtrack_stack: Vec<BacktrackFrame> = Vec::new();
+    let mut remaining_deps: RemainingDeps = Vec::new();
+    remaining_deps.extend(
+        try!(activate(&mut cx, registry, top, &top_method)));
+    loop {
+        // Retrieves the next dependency to try, from `remaining_deps`.
+        let (parent, platform, cur, dep, candidates, features) =
+            match remaining_deps.pop() {
+                None => break,
+                Some(mut deps_frame) => {
+                    let info =
+                        match deps_frame.remaining_siblings.next() {
+                            Some((cur, &(ref dep, ref candidates, ref features))) =>
+                                (deps_frame.parent.clone(), deps_frame.platform.clone(),
+                                 cur, dep.clone(),
+                                 candidates.clone(), features.clone()),
+                            None => {
+                                cx.visited.remove(&deps_frame.id);
+                                continue;
+                            }
+                        };
+                    remaining_deps.push(deps_frame);
+                    info
+                }
+            };
```

Comment:
I don't follow. Are you saying you want the `None` branch to come second or ...? This got a bunch simpler when I made `RcVecIter` auto-clone, so see if I accidentally put it in the right form.